src/Lib.P2P/PubSub/NotificationService.cs

```diff
@@ -0,0 +1,242 @@
+            DuplicateMesssagesReceived = 0;
+
+            // Listen to the routers.
+            foreach (var router in Routers)
+            {
+                router.MessageReceived += Router_MessageReceived;
+                await router.StartAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task StopAsync()
+        {
+            topicHandlers.Clear();
+
+            foreach (var router in Routers)
+            {
+                router.MessageReceived -= Router_MessageReceived;
+                await router.StopAsync();
+            }
+        }
+
+        /// <summary>
+        ///   Creates a message for the topic and data.
+        /// </summary>
+        /// <param name="topic">
+        ///   The topic name/id.
+        /// </param>
+        /// <param name="data">
+        ///   The payload of message.
+        /// </param>
+        /// <returns>
+        ///   A unique published message.
+        /// </returns>
+        /// <remarks>
+        ///   The <see cref="PublishedMessage.SequenceNumber"/> is a monitonically 
+        ///   increasing unsigned long.
+        /// </remarks>
+        public PublishedMessage CreateMessage(string topic, byte[] data)
+        {
+            var next = Interlocked.Increment(ref nextSequenceNumber);
+            var seqno = BitConverter.GetBytes(next);
+            if (BitConverter.IsLittleEndian) seqno = seqno.Reverse().ToArray();
+            return new PublishedMessage
+            {
+                Topics = new string[] {topic},
+                Sender = LocalPeer,
+                SequenceNumber = seqno,
+                DataBytes = data
+            };
+        }
+
+        /// <inheritdoc />
+        public Task<IEnumerable<string>> SubscribedTopicsAsync(CancellationToken cancel = default(CancellationToken))
+        {
+            var topics = topicHandlers.Values
+               .Select(t => t.Topic)
+               .Distinct();
+            return Task.FromResult(topics);
+        }
+
+        /// <inheritdoc />
+        public Task<IEnumerable<Peer>> PeersAsync(string topic = null,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            var peers = Routers
+               .SelectMany(r => r.InterestedPeers(topic))
+               .Distinct();
+            return Task.FromResult(peers);
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync(string topic, string message, CancellationToken cancel = default(CancellationToken))
+        {
+            return PublishAsync(topic, Encoding.UTF8.GetBytes(message), cancel);
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync(string topic, Stream message, CancellationToken cancel = default(CancellationToken))
+        {
+            using (var ms = new MemoryStream())
+            {
+#pragma warning disable VSTHRD103
+                message.CopyTo(ms);
+#pragma warning disable VSTHRD103 
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)